server/lib/device/device.getFeatureStates.js

```diff
@@ -0,0 +1,117 @@
+const { Op } = require('sequelize');
+const db = require('../../models');
+
+const DEFAULT_OPTIONS = {
+  skip: 0,
+  order_dir: 'ASC',
+  order_by: 'name',
+};
+
+/**
+ * @description Get list of device feature states
+ * @param {Object} [options] - Options of the query.
+ * @returns {Promise} List of device with device feature state.
+ * @example
+ * const devices = await gladys.device.getFeatureStates({
+ *  device_feature_selector: 'test-selector'
+ * });
+ */
+async function getFeatureStates(options) {
+  const optionsWithDefault = Object.assign({}, DEFAULT_OPTIONS, options);
+
+  // Build query for not feature_state info
+  const queryParams = {
+    include: [
+      {
+        model: db.DeviceFeature,
+        as: 'features',
+      },
+      {
+        model: db.Room,
+        as: 'room',
+      },
+    ],
+    offset: optionsWithDefault.skip,
+  };
+
+  // Fix attributes
+  if (optionsWithDefault.attributes_device && optionsWithDefault.attributes_device.length > 0) {
+    queryParams.attributes = optionsWithDefault.attributes_device;
+  }
+  if (optionsWithDefault.attributes_device_feature && optionsWithDefault.attributes_device_feature.length > 0) {
+    queryParams.include[0].attributes = optionsWithDefault.attributes_device_feature;
+  }
+  if (optionsWithDefault.attributes_device_room && optionsWithDefault.attributes_device_room.length > 0) {
+    queryParams.include[1].attributes = optionsWithDefault.attributes_device_room;
```

Comment:
Could you explained how this will be queried? it's a little bit weird & hard to understand

Reply:
this part is for reduce data up from database (reduce attribute in select query) so it's just to improve the performance of object serialization .

so I apply list of column I want get by object in join query, in this case I want get device like principal object and get feature and room with the minimum of column I need.

to be more flexible I pass this list of attribute in function parameter to authorize other developer to use this function with this limitation